### PR TITLE
[WasmFS] Simplify the results of `Directory::getEntries`

### PR DESCRIPTION
**Pull request by tlively**
Instead of forcing backends to materialize full `File` objects for the entries
when calling `getEntries`, only return the basic information needed to implement
`getdents`.

### CODE REVIEW

File: system/lib/wasmfs/syscalls.cpp [hunk 1]
```diff
@@ -728,20 +727,36 @@ long __syscall_getdents64(long fd, long dirp, long count) {
...
+    result->d_ino = entry.ino;
     result->d_off = bytesRead + sizeof(dirent);
     result->d_reclen = sizeof(dirent);
-    result->d_type =
-      curr.file->is<Directory>() ? DT_DIR : DT_REG; // TODO: add symlinks.
-    assert(curr.name.size() + 1 <= sizeof(result->d_name));
-    strcpy(result->d_name, curr.name.c_str());
+    switch (entry.kind) {
+      case File::UnknownKind:
+        result->d_type = DT_UNKNOWN;
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:743]: Fascinating, so this is allowed in the spec for the exact reason we want, which is to not do more work to figure it out eagerly...

**Review by kripken — APPROVED** [review]
Side note, inode ID handling may get "interesting" in the future, with lazy backends. They'd need to report some inode in getEntries, but then if they create a file then currently its inode is its pointer, which may be a different value... but maybe that is ok, if the inode is allowed to change before a file is actually opened and used?

File: system/lib/wasmfs/file.h [hunk 2]
```diff
@@ -30,7 +30,9 @@ const backend_t NullBackend = nullptr;
 
 class File : public std::enable_shared_from_this<File> {
 public:
-  enum FileKind { DataFileKind = 0, DirectoryKind, SymlinkKind };
+  enum FileKind { UnknownKind, DataFileKind, DirectoryKind, SymlinkKind };
```
Comment (sbc100) [inline system/lib/wasmfs/file.h:33]: Why not reuse DT_CHR and friends here?

Comment (tlively) [reply system/lib/wasmfs/file.h:33]: Mostly for readability, but also to make it more explicit that we only support a subset of the available file types.

**Comment by tlively** [comment]
> Side note, inode ID handling may get "interesting" in the future, with lazy backends. They'd need to report some inode in getEntries, but then if they create a file then currently its inode is its pointer, which may be a different value... but maybe that is ok, if the inode is allowed to change before a file is actually opened and used?

Yeah, technically that would be allowed because it's the same as someone replacing the files between the directory read and the file access. We may want to make `getIno` virtual and let backends override it in the future, though. Otherwise the inodes aren't really useful for their intended purpose, which is telling whether or not two files are actually the same file.

**Pull request merged**